platforms/jvm/testing-jvm/src/testFixtures/groovy/testengines/TestEnginesFixture.groovy

```diff
@@ -0,0 +1,103 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//file:noinspection unused
+
+package testengines
+
+import groovy.transform.SelfType
+import org.gradle.api.logging.configuration.ConsoleOutput
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.TestResources
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.integtests.fixtures.executer.GradleDistribution
+import org.gradle.integtests.fixtures.executer.GradleExecuter
+import org.gradle.integtests.fixtures.executer.IntegrationTestBuildContext
+import org.gradle.integtests.fixtures.executer.UnderDevelopmentGradleDistribution
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.gradle.util.internal.TextUtil
+
+/**
+ * Trait to provide custom JUnit Test Engines to integration tests.
+ * <p>
+ * Note that this trait can <strong>NOT BE SHARED</strong> by multiple test classes in the same project
+ * that share a class name prefix (e.g. MyClassRedTest and MyClassBlueTest) because of the way our testing
+ * infrastructure sets up test directories.  Both of these test classes would use the same engine copy-to
+ * dir, and thus interfere with each other.
+ */
+@SelfType(AbstractIntegrationSpec)
+trait TestEnginesFixture {
+    private static final String ENGINE_COPY_TO_DIR_NAME = "test-engine-build"
+
+    private static File engineBuildDir
+    private static String engineJarLibPath
+
+    abstract List<TestEngines> getEnginesToSetup()
+
+    def setupSpec() {
+        // Create a custom test directory provider to isolate the engine build as a sibling of the test method dirs
+        TestNameTestDirectoryProvider customTestDirectoryProvider = new TestNameTestDirectoryProvider.ParentDirectoryProvider(this.getClass())
+
+        // Copy required test engine source to the root of the test directory structure for this test class
+        TestResources resources = new TestResources(customTestDirectoryProvider, TestEngines.class, TestEngines.class)
+        assert resources.maybeCopy("shared")
+        getEnginesToSetup().forEach {
+            assert resources.maybeCopy(it.name)
+        }
+
+        // Switch to engine build directory for this setup
+        GradleDistribution distribution = new UnderDevelopmentGradleDistribution(IntegrationTestBuildContext.INSTANCE)
+        GradleExecuter engineBuilder = new GradleContextualExecuter(distribution, customTestDirectoryProvider, IntegrationTestBuildContext.INSTANCE)
+        engineBuildDir = customTestDirectoryProvider.testDirectory.file(ENGINE_COPY_TO_DIR_NAME)
+        engineBuilder.inDirectory(engineBuildDir)
+            .withRepositoryMirrors()
+            .withConsole(ConsoleOutput.Verbose)
+
+        // Build the test engine jar
+        engineBuilder.withTasks("build").run()
```

Comment:
I think `ArtifactBuilder` does something like this already

Reply:
Looking at ArtifactBuilder, it's about programatically building the jar file by file, not copying pre-defined source files to build.  I think I'd rather leave this separate, since the engine examples are more complex.